journal/src/test/java/io/zeebe/journal/file/SegmentedJournalTest.java

```diff
@@ -0,0 +1,376 @@
+
+  @Test
+  public void shouldBeEmpty() {
+    // when-then
+    assertThat(journal.isEmpty()).isTrue();
+  }
+
+  @Test
+  public void shouldNotBeEmpty() {
+    // given
+    journal.append(1, data);
+
+    // when-then
+    assertThat(journal.isEmpty()).isFalse();
+  }
+
+  @Test
+  public void shouldAppendData() {
+    // given
+    final var recordAppended = journal.append(1, data);
+    assertThat(recordAppended.index()).isEqualTo(1);
+
+    // when
+    final var recordRead = journal.openReader().next();
+
+    // then
+    assertThat(recordAppended.index()).isEqualTo(recordRead.index());
+    assertThat(recordAppended.asqn()).isEqualTo(recordRead.asqn());
+    assertThat(recordAppended.checksum()).isEqualTo(recordRead.checksum());
+  }
+
+  @Test
+  public void shouldAppendMultipleRecords() {
+    for (int i = 0; i < 10; i++) {
+      final var recordAppended = journal.append(i + 10, data);
+      assertThat(recordAppended.index()).isEqualTo(i + 1);
+    }
+
+    final var reader = journal.openReader();
+    for (int i = 0; i < 10; i++) {
+      assertThat(reader.hasNext()).isTrue();
+      final var recordRead = reader.next();
+      assertThat(recordRead.index()).isEqualTo(i + 1);
+      final byte[] data = new byte[recordRead.data().capacity()];
+      recordRead.data().getBytes(0, data);
+      assertThat(recordRead.asqn()).isEqualTo(i + 10);
+      assertThat(data).containsExactly(entry);
+    }
+  }
+
+  @Test
+  public void shouldAppendAndReadMultipleRecords() {
+    final var reader = journal.openReader();
+    for (int i = 0; i < 10; i++) {
+      // given
+      entry = ("TestData" + i).getBytes();
+      data.wrap(entry);
+
+      // when
+      final var recordAppended = journal.append(i + 10, data);
+      assertThat(recordAppended.index()).isEqualTo(i + 1);
+
+      // then
+      assertThat(reader.hasNext()).isTrue();
+      final var recordRead = reader.next();
+      assertThat(recordRead.index()).isEqualTo(i + 1);
+      final byte[] data = new byte[recordRead.data().capacity()];
+      recordRead.data().getBytes(0, data);
+      assertThat(recordRead.asqn()).isEqualTo(i + 10);
+      assertThat(data).containsExactly(entry);
+    }
+  }
+
+  @Test
+  public void shouldReset() {
+    // given
+    long asqn = 1;
+    assertThat(journal.getLastIndex()).isEqualTo(0);
+    journal.append(asqn++, data);
+    journal.append(asqn++, data);
+
+    // when
+    journal.reset(2);
+
+    // then
+    assertThat(journal.getLastIndex()).isEqualTo(1);
+    final var record = journal.append(asqn++, data);
+    assertThat(record.index()).isEqualTo(2);
+  }
+
+  @Test
+  public void shouldResetWhileReading() {
+    // given
+    final var reader = journal.openReader();
+    long asqn = 1;
+    assertThat(journal.getLastIndex()).isEqualTo(0);
+    journal.append(asqn++, data);
+    journal.append(asqn++, data);
+    final var record1 = reader.next();
+    assertThat(record1.index()).isEqualTo(1);
+
+    // when
+    journal.reset(2);
```

Comment:
This is confusing :smile: I would expect that last index is 2 ? :thinking:

Reply:
Yes. Last index is 2. reset(2) means that nextIndex will be 2. So the record at index 2 will be cleared.

Reply:
So it is actually `reset to 1`

Reply:
No. It is reset to 2. The doc of reset(nextIndex) says `Delete all records in the journal and reset the next index to nextIndex. The following calls to Journal#append will append at index nextIndex.`